### PR TITLE
Charlesmchen/crop and scale avatar

### PR DESCRIPTION
**Pull request by charlesmchen**
Affects profile and group avatars.  

In the long run, we should combine this with the confirmation view, but I'm not sure it's worth the trouble now.

PTAL @michaelkirk 

img_0037 1
img_0038 1

### CODE REVIEW

File: Signal/src/ViewControllers/CropScaleImageViewController.swift [hunk 1]
```diff
@@ -0,0 +1,501 @@
...
+            srcTranslation
+                = srcTranslationAtPanStart
+            break
+        }
+
+        updateImageLayout()
+    }
+
+    func handleDoubleTap(sender: UIPanGestureRecognizer) {
+        if (sender.state == .recognized) {
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/CropScaleImageViewController.swift]: TIOLI:

1. I was expecting this to zoom in.
2. iMessage avatar editing doesn't seem to do anything with a double tap.

File: Signal/src/ViewControllers/CropScaleImageViewController.swift [hunk 2]
```diff
@@ -0,0 +1,501 @@
...
+        let hasAlpha = false
+        let dstScale: CGFloat = 1.0 // The size is specified in pixels, not in points.
+        UIGraphicsBeginImageContextWithOptions(dstSizePixels, !hasAlpha, dstScale)
+
+        let imageViewFrame = imageRenderRect(forDstSize:dstSizePixels)
+        srcImage.draw(in:imageViewFrame)
+
+        let scaledImage = UIGraphicsGetImageFromCurrentImageContext()
+        if scaledImage == nil {
+            Logger.error("\(TAG) could not generate dst image.")
```
Comment (michaelkirk) [inline Signal/src/ViewControllers/CropScaleImageViewController.swift]: OWSFail?

**Comment by charlesmchen** [comment]
> just nits, plus you got my earlier design feedback.

I've added a mask to crop to circle.  I'm adding a nice-to-have card to show the cropped image content.

File: Signal/src/ViewControllers/CropScaleImageViewController.swift [hunk 3]
```diff
@@ -0,0 +1,501 @@
+//
+//  Copyright (c) 2017 Open Whisper Systems. All rights reserved.
+//
+
+import Foundation
+import MediaPlayer
+
+class OWSLayerView: UIView {
+    var layoutCallback : (() -> Void)?
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/CropScaleImageViewController.swift]: Done, better.

File: Signal/src/ViewControllers/CropScaleImageViewController.swift [hunk 4]
```diff
@@ -0,0 +1,501 @@
...
+
+class CropScaleImageViewController: OWSViewController {
+
+    let TAG = "[CropScaleImageViewController]"
+
+    // MARK: Properties
+
+    let srcImage: UIImage
+
+    var successCompletion: ((UIImage) -> Void)?
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/CropScaleImageViewController.swift]: Done.

File: Signal/src/ViewControllers/CropScaleImageViewController.swift [hunk 5]
```diff
@@ -0,0 +1,501 @@
...
+
+    var imageView: UIView?
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/CropScaleImageViewController.swift]: Ah interesting.

File: Signal/src/ViewControllers/CropScaleImageViewController.swift [hunk 6]
```diff
@@ -0,0 +1,501 @@
...
+    var imageScale: CGFloat = 1.0
+
+    // This represents the current translation from the
+    // upper-left corner of the src image to the upper-left
+    // corner of the crop region in src image point coordinates.
+    var srcTranslation: CGPoint = CGPoint.zero
+
+    // MARK: Initializers
+
+    @available(*, unavailable, message:"use attachment: constructor instead.")
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/CropScaleImageViewController.swift]: Done.

File: Signal/src/ViewControllers/CropScaleImageViewController.swift [hunk 7]
```diff
@@ -0,0 +1,501 @@
...
+    private func configureCropAndScale() {
+        // Size of bounding box that reflects the target aspect ratio, whose longer side = 1.
+        let unitSquareHeight: CGFloat = (dstAspectRatio >= 1.0 ? 1.0 : 1.0 / dstAspectRatio)
+        let unitSquareWidth: CGFloat = (dstAspectRatio >= 1.0 ? dstAspectRatio * unitSquareHeight : 1.0)
+        let unitSquareSize = CGSize(width: unitSquareWidth, height: unitSquareHeight)
+
+        srcImageSizePoints = srcImage.size
+        guard
+            (srcImageSizePoints.width > 0 && srcImageSizePoints.height > 0) else {
+                return
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/CropScaleImageViewController.swift]: I believe most image formats are perfectly happy to let you create a gif, png, jpg of zero width and/or height.  The user could pick such an image from their camera roll.  I'm trying to be super defensive and avoid divide by zeroes and similar crashes in production. If this happened in prod, the view would look empty and hitting done with generate a blank (undefined?) avatar image of the correct size.  I guess we could fill the avatar with white or black in this case but I'm mostly worried about being crash-proof.

File: Signal/src/ViewControllers/CropScaleImageViewController.swift [hunk 8]
```diff
@@ -0,0 +1,501 @@
...
+        updateImageLayout()
+    }
+
+    override func viewDidAppear(_ animated: Bool) {
+        super.viewDidAppear(animated)
+
+        updateImageLayout()
+    }
+
+    private func defaultCropFramePoints(imageSizePoints: CGSize, viewSizePoints: CGSize) -> (CGRect) {
```
[the remark this answers is not in the document]
Comment (charlesmchen) [reply Signal/src/ViewControllers/CropScaleImageViewController.swift]: Done.

File: Signal/src/ViewControllers/CropScaleImageViewController.swift (hunk 1 above)
Comment (charlesmchen) [reply Signal/src/ViewControllers/CropScaleImageViewController.swift]: okay, I'll skip double tap.

File: Signal/src/ViewControllers/CropScaleImageViewController.swift (hunk 2 above)
Comment (charlesmchen) [reply Signal/src/ViewControllers/CropScaleImageViewController.swift]: Done.

**Comment by charlesmchen** [comment]
I've added a bunch more comments.

**Pull request merged**